pkg/monitoring/prometheusagent/service.go

```diff
@@ -0,0 +1,232 @@
+	objectKey := client.ObjectKey{
+		Name:      getPrometheusAgentRemoteWriteConfigName(cluster),
+		Namespace: cluster.GetNamespace(),
+	}
+
+	current := &corev1.ConfigMap{}
+	// Get the current configmap if it exists.
+	err := pas.Client.Get(ctx, objectKey, current)
+	if apierrors.IsNotFound(err) {
+		configMap, err := pas.buildRemoteWriteConfig(ctx, cluster, logger, defaultShards)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		err = pas.Client.Create(ctx, configMap)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		return nil
+	} else if err != nil {
+		return errors.WithStack(err)
+	}
+
+	currentShards, err := readCurrentShardsFromConfig(*current)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	desired, err := pas.buildRemoteWriteConfig(ctx, cluster, logger, currentShards)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	if !reflect.DeepEqual(current.Data, desired.Data) || !reflect.DeepEqual(current.Finalizers, desired.Finalizers) {
+		err = pas.Client.Update(ctx, desired)
+		if err != nil {
+			logger.Info("could not update prometheus agent remote write configuration")
+			return errors.WithStack(err)
+		}
+	}
+	return nil
+}
+
+func (pas PrometheusAgentService) createOrUpdateSecret(ctx context.Context,
+	cluster *clusterv1.Cluster, logger logr.Logger) error {
+	objectKey := client.ObjectKey{
+		Name:      getPrometheusAgentRemoteWriteSecretName(cluster),
+		Namespace: cluster.GetNamespace(),
+	}
+
+	current := &corev1.Secret{}
+	// Get the current secret if it exists.
+	err := pas.Client.Get(ctx, objectKey, current)
+	if apierrors.IsNotFound(err) {
+		logger.Info("generating password for the prometheus agent")
+		password, err := pas.PasswordManager.GeneratePassword(32)
+		if err != nil {
+			logger.Error(err, "failed to generate the prometheus agent password")
+			return errors.WithStack(err)
+		}
+		logger.Info("generated password for the prometheus agent")
+
+		secret, err := pas.buildRemoteWriteSecret(cluster, password)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		err = pas.Client.Create(ctx, secret)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		return nil
+	} else if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// As it takes a long time to apply the new password to the agent due to a built-in delay in the app-platform,
+	// we keep the already generated remote write password.
+	password, err := readRemoteWritePasswordFromSecret(*current)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	desired, err := pas.buildRemoteWriteSecret(cluster, password)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	if !reflect.DeepEqual(current.Data, desired.Data) || !reflect.DeepEqual(current.Finalizers, desired.Finalizers) {
+		err = pas.Client.Update(ctx, desired)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+	}
+
+	return nil
+}
+
+func (pas *PrometheusAgentService) DeleteRemoteWriteConfig(ctx context.Context, cluster *clusterv1.Cluster) error {
```

Comment:
I'd rename it `DeleteRemoteWriteConfigAndSecret`

Reply:
What do you think about DeleteRemoteWriteConfiguration?

Reply:
Done